### PR TITLE
10007 test row filtering on null column

### PR DESCRIPTION
**Pull request by crpeter**

### CODE REVIEW

File: core/trino-main/src/test/java/io/trino/connector/MockConnector.java [hunk 1]
```diff
@@ -123,6 +123,7 @@
     private final Optional<ConnectorAccessControl> accessControl;
     private final Function<SchemaTableName, List<List<?>>> data;
     private final Set<Procedure> procedures;
+    private final Boolean allowMissingColumnsOnInsert;
```
Comment (kokosing) [inline core/trino-main/src/test/java/io/trino/connector/MockConnector.java:126]: `boolean`, use primitive

File: core/trino-main/src/test/java/io/trino/connector/MockConnectorFactory.java [hunk 2]
```diff
@@ -470,6 +474,11 @@ public Builder withColumnMask(BiFunction<SchemaTableName, String, ViewExpression
             return this;
         }
 
+        public Builder withallowMissingColumnsOnInsert(boolean allowMissingColumnsOnInsert) {
```
Comment (kokosing) [inline core/trino-main/src/test/java/io/trino/connector/MockConnectorFactory.java:477]: put `{` to new line

File: core/trino-main/src/test/java/io/trino/sql/query/TestRowFilter.java [hunk 3]
```diff
@@ -554,4 +562,24 @@ public void testRowFilterOnHiddenColumn()
                 .skippingTypesCheck()
                 .matches("SELECT BIGINT '25'");
     }
+
+    @Test
+    public void testRowFilterOnOptionalColumn() {
```
Comment (kokosing) [inline core/trino-main/src/test/java/io/trino/sql/query/TestRowFilter.java:567]: formatting: `{` to new line

File: core/trino-main/src/test/java/io/trino/sql/query/TestRowFilter.java [hunk 4]
```diff
@@ -554,4 +562,24 @@ public void testRowFilterOnHiddenColumn()
...
+        accessControl.reset();
+        accessControl.rowFilter(
+                new QualifiedObjectName(MOCK_CATALOG, "tiny", "nation_with_optional_column"),
+                USER,
+                new ViewExpression(USER, Optional.empty(), Optional.empty(), "optional is null"));
+
+        assertThatThrownBy(() -> assertions.query("INSERT INTO mock.tiny.nation_with_optional_column VALUES (101, 'POLAND', 0, 'No comment', 'some string')"))
```
Comment (kokosing) [inline core/trino-main/src/test/java/io/trino/sql/query/TestRowFilter.java:574]: What is a default value? If it is `null` then it should pass.

Comment (crpeter) [reply core/trino-main/src/test/java/io/trino/sql/query/TestRowFilter.java:574]: Could you clarify this?

File: core/trino-main/src/main/java/io/trino/metadata/Metadata.java [hunk 5]
```diff
@@ -306,7 +306,7 @@
     /**
      * @return whether connector handles missing columns during insert
      */
-    boolean supportsMissingColumnsOnInsert(Session session, TableHandle tableHandle);
+    boolean supportsMissingColumnsOnInsert(Session session, TableHandle tableHandle, QualifiedObjectName qualifiedObjectName);
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/metadata/Metadata.java:309]: Adding a test should not modify the existing product code? `TableHandle` should already have information about its name, so why you are passing `qualifiedObjectName`?

File: plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixMetadata.java [hunk 6]
```diff
@@ -162,7 +162,7 @@ public ConnectorOutputTableHandle beginCreateTable(ConnectorSession session, Con
     }
 
     @Override
-    public boolean supportsMissingColumnsOnInsert()
+    public boolean supportsMissingColumnsOnInsert(SchemaTableName viewName)
```
Comment (kokosing) [inline plugin/trino-phoenix/src/main/java/io/trino/plugin/phoenix/PhoenixMetadata.java:165]: Please undo SPI changes.

File: core/trino-main/src/test/java/io/trino/sql/query/TestRowFilter.java [hunk 7]
```diff
@@ -98,8 +102,17 @@ public void init()
                     if (schemaTableName.equals(new SchemaTableName("tiny", "nation_with_hidden_column"))) {
                         return TPCH_WITH_HIDDEN_COLUMN_DATA;
                     }
+                    if (schemaTableName.equals(new SchemaTableName("tiny", "nation_with_optional_column"))) {
+                        return TPCH_NATION_DATA;
+                    }
                     throw new UnsupportedOperationException();
                 })
+                .withAllowMissingColumnsOnInsert(schemaTableName -> {
```
Comment (kokosing) [inline core/trino-main/src/test/java/io/trino/sql/query/TestRowFilter.java:110]: Instead of modyfying SPI, I think it would be better to setup another catalog in this test with a different MockConnectorFactory setup.

File: core/trino-main/src/test/java/io/trino/connector/MockConnectorEntities.java [hunk 8]
```diff
@@ -39,6 +39,11 @@ private MockConnectorEntities() {}
             .add(ColumnMetadata.builder().setName("$hidden").setType(BigintType.BIGINT).setHidden(true).build())
             .build();
 
+    public static final List<ColumnMetadata> TPCH_NATION_WITH_OPTIONAL_COLUMN = ImmutableList.<ColumnMetadata>builder()
+            .addAll(TPCH_NATION_SCHEMA)
+            .add(ColumnMetadata.builder().setName("debt").setType(BigintType.BIGINT).setNullable(true).build())
```
Comment (kokosing) [inline core/trino-main/src/test/java/io/trino/connector/MockConnectorEntities.java:44]: use `optional_column` instead.

File: core/trino-main/src/test/java/io/trino/sql/query/TestRowFilter.java [hunk 9]
```diff
@@ -554,4 +567,29 @@ public void testRowFilterOnHiddenColumn()
...
+                new QualifiedObjectName(MOCK_CATALOG, "tiny", "nation_with_optional_column"),
+                USER,
+                new ViewExpression(USER, Optional.empty(), Optional.empty(), "debt > 0"));
+
+        assertions.query("INSERT INTO mock.tiny.nation_with_optional_column(nationkey, name, regionkey, comment, debt) VALUES (0, 'POLAND', 0, 'No comment', 1)")
+                .assertThat()
+                .skippingTypesCheck()
+                .matches("VALUES BIGINT '1'");
+
+        assertThatThrownBy(() -> assertions.query("INSERT INTO mock.tiny.nation_with_optional_column(nationkey, name, regionkey, comment, debt) VALUES (0, 'POLAND', 0, 'No comment', 0)"))
```
Comment (kokosing) [inline core/trino-main/src/test/java/io/trino/sql/query/TestRowFilter.java:591]: add a test when filter is passing and optional column is used.

File: core/trino-main/src/test/java/io/trino/sql/query/TestRowFilter.java [hunk 10]
```diff
@@ -554,4 +596,33 @@ public void testRowFilterOnHiddenColumn()
                 .skippingTypesCheck()
                 .matches("SELECT BIGINT '25'");
     }
+
+    @Test
+    public void testRowFilterOnOptionalColumn()
+    {
+        accessControlMissingColumns.reset();
+
+        assertionsMissingColumns.query("INSERT INTO mockmissingcolumns.tiny.nation_with_optional_column(nationkey, name, regionkey, comment) VALUES (0, 'POLAND', 0, 'No comment')")
```
Comment (kokosing) [inline core/trino-main/src/test/java/io/trino/sql/query/TestRowFilter.java:605]: please extract this assertion to separate test method

File: core/trino-main/src/test/java/io/trino/sql/query/TestRowFilter.java [hunk 11]
```diff
@@ -554,4 +596,33 @@ public void testRowFilterOnHiddenColumn()
                 .skippingTypesCheck()
                 .matches("SELECT BIGINT '25'");
     }
+
+    @Test
+    public void testRowFilterOnOptionalColumn()
+    {
+        accessControlMissingColumns.reset();
```
Comment (kokosing) [inline core/trino-main/src/test/java/io/trino/sql/query/TestRowFilter.java:603]: How about extracting this to `@BeforeMethod` (please do it as separate commit).?
FYI Please follow: https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#git-merge-strategy

**Review by kokosing — COMMENTED** [review]
Looks nice. Thank you!

Please make sure you singed CLA and it processed properly. Also please follwo https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#git-merge-strategy and prepare your PR to be merged.

**Comment by crpeter** [comment]
> Looks nice. Thank you!
> 
> Please make sure you singed CLA and it processed properly. Also please follwo https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#git-merge-strategy and prepare your PR to be merged.

I submitted this to cla@trino.io on Dec 14, 2021
